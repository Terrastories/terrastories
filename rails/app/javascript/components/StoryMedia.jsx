import React, { PureComponent } from "react";
import PropTypes from 'prop-types';

class StoryMedia extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      explicitVideoHeight: null
    }
  }

  static propTypes = {
    file: PropTypes.object,
    doBustCache: PropTypes.func
  };

  renderAudio() {
    const { file } = this.props;
    return (
      <audio id={`audio-player${file.blob.id}`} controls>
        <source src={file.url} type={file.blob.content_type} />
      </audio>
    );
  }

  renderVideo() {
    const { file } = this.props;
    const { explicitVideoHeight } = this.state;
    return (
      <video
        id={`video-player${file.blob.id}`}
        className="video-player"
        height={explicitVideoHeight}
        playsInline
        controls
        disablepictureinpicture='true'       
        controlsList='nodownload'
        key={file.url}
        ref="video"
      >
        <source src={file.url} type={file.blob.content_type} />
      </video>
    );
  }

  render() {
    return this.props.file.blob.content_type.indexOf("video") !== -1
      ? this.renderVideo()
      : this.renderAudio();
  }

  componentDidMount() {
    const { doBustCache } = this.props;
    const video = this.refs.video;

    if (!!video === true) {
      video.addEventListener('loadeddata', (event) => {
        const aspectRatio = video.videoWidth / video.videoHeight;
        const newExplicitVideoHeight = video.offsetWidth / aspectRatio;
        this.setState({ explicitVideoHeight: newExplicitVideoHeight });
        doBustCache();
      });
      video.addEventListener('play', (event) => {
        window.pauseAllVideos(event.target);
      }, true);
    }
  }
}

export default StoryMedia;
