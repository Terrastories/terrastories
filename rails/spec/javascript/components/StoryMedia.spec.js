import React from 'react';
import { shallow } from 'enzyme';
import StoryMedia from 'components/StoryMedia';
it('should render correctly in "image" mode', () => {
    const props = {
        file: {
            url: '',
            blob: {
                id: 2,
                content_type: 'awesome'
            }
        }
    }
    const component = shallow(<StoryMedia {...props} />);

    expect(component).toMatchSnapshot();
    expect(component.find('img').exists()).toBe(true)
});