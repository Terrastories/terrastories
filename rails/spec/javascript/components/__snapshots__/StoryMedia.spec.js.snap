// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly in "image" mode 1`] = `
<img
  className="img-player"
  id="img2"
  src=""
  width="80%"
/>
`;
